server/src/main/java/org/opensearch/action/search/TransportDeletePITAction.java

```diff
@@ -0,0 +1,172 @@
+                searchTransportService.sendDeleteAllPitContexts(connection, groupedActionListener);
+            } catch (Exception e) {
+                groupedActionListener.onFailure(e);
+            }
+        }
+    }
+
+    /**
+     * Delete list of pits, return success if all reader contexts are deleted ( or not found ).
+     */
+    void deletePits(List<SearchContextIdForNode> contexts, ActionListener<Integer> listener) {
+        final StepListener<BiFunction<String, String, DiscoveryNode>> lookupListener = getLookupListener(contexts);
+        lookupListener.whenComplete(nodeLookup -> {
+            final GroupedActionListener<Boolean> groupedListener = new GroupedActionListener<>(
+                ActionListener.delegateFailure(
+                    listener,
+                    (l, result) -> l.onResponse(Math.toIntExact(result.stream().filter(r -> r).count()))
+                ),
+                contexts.size()
+            );
+
+            for (SearchContextIdForNode contextId : contexts) {
+                final DiscoveryNode node = nodeLookup.apply(contextId.getClusterAlias(), contextId.getNode());
+                if (node == null) {
+                    groupedListener.onFailure(new OpenSearchException("node not connected"));
+                } else {
+                    try {
+                        final Transport.Connection connection = searchTransportService.getConnection(contextId.getClusterAlias(), node);
+                        searchTransportService.sendPitFreeContext(
+                            connection,
+                            contextId.getSearchContextId(),
+                            ActionListener.wrap(r -> groupedListener.onResponse(r.isFreed()), e -> groupedListener.onResponse(false))
+                        );
+                    } catch (Exception e) {
+                        logger.debug("Delete PIT failed ", e);
+                        groupedListener.onResponse(false);
+                    }
+                }
+            }
+        }, listener::onFailure);
+    }
+
+    private StepListener<BiFunction<String, String, DiscoveryNode>> getLookupListener(List<SearchContextIdForNode> contexts) {
+        final StepListener<BiFunction<String, String, DiscoveryNode>> lookupListener = new StepListener<>();
+        final Set<String> clusters = contexts.stream()
+            .filter(ctx -> Strings.isEmpty(ctx.getClusterAlias()) == false)
+            .map(SearchContextIdForNode::getClusterAlias)
+            .collect(Collectors.toSet());
+        if (clusters.isEmpty() == false) {
+            searchTransportService.getRemoteClusterService().collectNodes(clusters, lookupListener);
+        } else {
+            lookupListener.onResponse((cluster, nodeId) -> clusterService.state().getNodes().get(nodeId));
+        }
+        return lookupListener;
+    }
+
+    private ActionListener<DeletePITResponse> getGroupedListener(ActionListener<DeletePITResponse> deletePitListener, int size) {
```

Comment:
why a separate method for this if its used only once and is private hence not even being called in unit tests?